sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/Analyzer.scala

```diff
@@ -212,31 +212,96 @@ class Analyzer(
   )
 
   /**
-   * Analyze cte definitions and substitute child plan with analyzed cte definitions.
+   * Analyze WITH nodes and substitute child plan with CTE definitions.
    */
   object CTESubstitution extends Rule[LogicalPlan] {
-    def apply(plan: LogicalPlan): LogicalPlan = plan.resolveOperatorsUp {
-      case With(child, relations) =>
-        // substitute CTE expressions right-to-left to resolve references to previous CTEs:
-        // with a as (select * from t), b as (select * from a) select * from b
-        relations.foldRight(child) {
-          case ((cteName, ctePlan), currentPlan) =>
-            substituteCTE(currentPlan, cteName, ctePlan)
-        }
-      case other => other
-    }
+    def apply(plan: LogicalPlan): LogicalPlan = traverse(plan, false)
```

Comment:
I feel the current logic is a little hard to understand because of the legacy mode.
I think the existing logic for WITH is pretty simple, so how about keeping the existing one and adding a new logic for subqueries like this?
```
  object CTESubstitution extends Rule[LogicalPlan] {
    def apply(plan: LogicalPlan): LogicalPlan = if (!SQLConf.get.legacyCTESubstitutionEnabled) {
      traverseAndsubstituteCTE(plan)
    } else {
      legacyTraverseAndsubstituteCTE(plan)
    }
```

Reply:
Thanks @maropu for the review! I changed as you suggested.